### PR TITLE
pkg/promtail: IETF Syslog (RFC5424) Support

### LINKED ISSUE #153: Support syslog-ng
https://twitter.com/PCzanik/status/1074996519638568961

https://www.syslog-ng.com/technical-documents/doc/syslog-ng-open-source-edition/3.19/administration-guide/36#TOPIC-1094599

https://www.syslog-ng.com/community/b/blog/posts/bulk-mode-message-sending-to-elasticsearch-with-syslog-ng-http-destination

### LINKED ISSUE #935: A way to ingest syslog messages over the network and into Loki
**Is your feature request related to a problem? Please describe.**

There are many devices such as firewalls/network appliances/switches etc that don't write log files or run on architectures too difficult to compile promtail for, however many of them support shipping syslogs (as well as pretty much all *nix servers in general).  It would be great to have a way to receive these syslogs and get them into Loki

**Describe the solution you'd like**

I think this needs some discussion, should Promtail be updated to receive these syslogs and then send them to Loki? Or another tool created? Or Loki updated to receive them directly?

### PR DESCRIPTION
**Pull request by bastjan**
**What this PR does / why we need it**:

Allow promtail to ingest [IETF syslog (RFC5424)](https://tools.ietf.org/html/rfc5424).

**Limitations**:

- No remote time support (tried it but got out-of-order errors due to data races across the network)
- No UDP/ No [BSD syslog (RFC3164)](https://tools.ietf.org/html/rfc3164).

**Performance**:

Tested on my somewhat slow machine (Intel Core i5-3210M). Used the `loggen` tool from syslog-ng with 200 parallel log streams. Minimal relabeling.

<details>
<summary>Configuration</summary>

```yaml
server:
  http_listen_port: 9080
  grpc_listen_port: 0

positions:
  filename: /tmp/positions.yaml

clients:
  - url: http://localhost:3100/loki/api/v1/push

scrape_configs:
- job_name: syslog
  syslog:
    listen_address: localhost:1514
    labels:
      job: "syslog"
  relabel_configs:
  - source_labels: ['__syslog_message_hostname']
    target_label: 'hostname'
```

</details>

```sh
> loggen --syslog-proto --active-connections 200 --rate 10000 --stream localhost 1514
average rate = 78671.04 msg/sec, count=2124253, time=27.001, (average) msg size=260, bandwidth=19963.56 kB/sec
```

**Which issue(s) this PR fixes**:
Fixes #153, Fixes #935 

**Special notes for your reviewer**:

~~Currently tracks the `develop` branch of `github.com/influxdata/go-syslog` since the `v2.0.0` releases go mod support is broken.~~

**Checklist**
- [x] Documentation added
- [x] Tests updated

### CODE REVIEW

**Comment by rfratto** [comment]
This is great, and something we definitely want in Promtail! I talked with the team and it seems the UDP limitation is the first thing we might want to get resolved since it's so common in the syslog world. 

cc @slim-bean

**Comment by bastjan** [comment]
> This is great, and something we definitely want in Promtail! I talked with the team and it seems the UDP limitation is the first thing we might want to get resolved since it's so common in the syslog world.

Great to hear! I'm planning to bring this merge request to a mergeable state in the next two days.

I would really appreciate a review of the code/concept before adding more features.

I never implemented an UDP server before in go but it does not look too complicated... I think I can do an initial version over the next two weeks.

**Comment by bastjan** [comment]
I used this feature for a while now in our staging environment with syslog-ng in front for UDP/TLS. Works great with an average of around 4400 lines/s (90% of the traffic is simulated). Memory usage hovers around 75MB.

**Comment by bastjan** [comment]
After using loki and promtail some more I'm wondering if the syslog parser should move to the `pipeline` stage. 

Something like this:
```yaml
scrape_configs:
- job_name: syslog
  syslog:
    listen_address: 0.0.0.0:1514
    labels:
      job: "syslog"
  entry_parser: syslog-rfc5452
  entry_parser_options:
    label_structured_data: yes
```
or
```yaml
scrape_configs:
- job_name: syslog
  syslog:
    listen_address: 0.0.0.0:1514
    labels:
      job: "syslog"
  entry_parser: raw
  pipeline_stages:
  - syslog-rfc5452:
      label_structured_data: yes
  - timestamp:
      source: timetamp
      format: 2006-01-02T15:04:05.999999Z07:00 # (RFC3339Micro)
```

I haven't used promtail enough to form an opinion on it.

RFC5452 is quite static and the spec is usually correctly implemented. So I don't feel like there is any advantage in flexibility.  RFC3164 (BSD Syslog) on the other hand is rarely correctly implemented, the format varies a lot, and I wonder if the parsing it should just be a pipeline stage.

**Comment by cyriltovena** [comment]
I don’t know much about syslog specifications (/cc @slim-bean ) but if there’s is no clear advantage I would go for the first config example.

I think it’s also better to keep the syslog configuration specific to the syslog scrape config.

I’m wondering if the use case of ingesting multiple specifications from one instance of promtail should be covered. Although we could run 2 instances with different config.

**Comment by bastjan** [comment]
> I’m wondering if the use case of ingesting multiple specifications from one instance of promtail should be covered. Although we could run 2 instances with different config.

You could always have two or more scrape configs.

A nice feature to have is support for both specifications on the same listener. Syslog-NG does not support this natively, and we had some problems because not every appliance does support setting a non-standard port. I've written a small syslog "router" at work that accepts both protocols and routes them to the correct Syslog-NG listener.

**Comment by pstibrany** [comment]
Thank you for your continuous improvements on this PR. It is not yet clear to us, whether syslog receiving should be part of Promtail or not. Syslog ecosystem is quite big, and there may be better alternatives (eg. push logs from syslog-ng / rsyslog directly through their http or program outputs, or perhaps even adding direct support there). That's not to say that we don't want this contribution, but we're still investigating what the best way forward is.

**Comment by pstibrany** [comment]
> I used this feature for a while now in our staging environment with syslog-ng in front for UDP/TLS.

Btw, since you're already running syslog-ng, how would direct exporter for syslog-ng to Loki work for you? I mean, without using promtail. I understand that you would lose some of the power of Promtail, on the other hand, direct export from syslog-ng would not lose logs if Promtail was down.

**Comment by bastjan** [comment]
> Btw, since you're already running syslog-ng, how would direct exporter for syslog-ng to Loki work for you?

That would work great, but a C syslog-ng plugin won't arrive in the mainstream distros for a while, and a python plugin is hard to package and deploy.

Having at least IETF syslog over TCP makes the integration of rsyslog and syslog-ng a breeze without writing plugins for both (and every other new product).

<details>
<summary>syslog-ng</summary>

```
destination d_loki {
  syslog("localhost" transport("tcp") port(6514));
};
```
</details>

<details>
<summary>rsyslog</summary>

```
action(type="omfwd" protocol="tcp" port="6514" Template="RSYSLOG_SyslogProtocol23Format")
```
</details>

I haven't tried to use the http output plugin of syslog-ng for loki yet, but the config will get complicated fast.

> direct export from syslog-ng would not lose logs if Promtail was down

For more complicated setups there can still be a syslog-ng/rsyslog in front with the "reliable" parameter set.

**Comment by pstibrany** [comment]
> I haven't tried to use the http output plugin of syslog-ng for loki yet, but the config will get complicated fast.

I looked at the documentation earlier today, and while basic http integration would not be that difficult, problem is that it would use one request per message. Syslog-ng http output also supports batching, but it would be difficult to massage output to what Loki currently expects.

**Comment by bastjan** [comment]
> I looked at the documentation earlier today, and while basic http integration would not be that difficult, problem is that it would use one request per message. Syslog-ng http output also supports batching, but it would be difficult to massage output to what Loki currently expects.

Yeah that's what I expected. I created a python plugin prototype some time ago which uses the provided gRPC definitions <https://gist.github.com/bastjan/cc8f25b6ea7395689e2e110836e312ce>. I can provide the full repo but have to strip company stuff out before...

**Comment by slim-bean** [comment]
Sorry I'm late to the party here.  I'm in favor of getting this added to promtail, where are we at with this, what do we need to get this merged?

I looked previously at the example config and would like to point out the `entry_parser` key has been deprecated and we should avoid using it, maybe we just need another key name for this.

File: pkg/promtail/targets/syslogtarget.go [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+				lb.Set(key, value)
+			}
+		}
+	}
+
+	processed := relabel.Process(lb.Labels(), t.relabelConfig...)
+
+	filtered := make(model.LabelSet)
+	for _, lbl := range processed {
+		if len(lbl.Name) >= 2 && lbl.Name[0:2] == "__" {
```
Comment (pstibrany) [inline pkg/promtail/targets/syslogtarget.go]:
```suggestion
		if strings.HasPrefix(lbl.Name, "__") {
```

File: pkg/promtail/targets/syslogtarget.go [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+func (t *SyslogTarget) Details() interface{} {
+	return map[string]string{}
+}
+
+// Stop shuts down the SyslogTarget.
+func (t *SyslogTarget) Stop() error {
+	close(t.shutdown)
+	err := t.listener.Close()
+	t.connectionsClosed.Wait()
+	close(t.messages)
```
Comment (pstibrany) [inline pkg/promtail/targets/syslogtarget.go]: Why is this needed?

File: pkg/promtail/targets/syslogtarget.go [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+	done := make(chan struct{})
+	defer close(done)
+	go func() {
+		select {
+		case <-done:
+		case <-t.shutdown:
+		}
+		c.Close()
+	}()
+
```
Comment (pstibrany) [inline pkg/promtail/targets/syslogtarget.go]: When using contexts, this would be replaced with

```
	handlerCtx, cancel := context.WithCancel(t.ctx)
	defer cancel()
	go func() {
		<-handlerCtx.Done()
		_ = c.Close()
	}()
```

(assuming `t.ctx` is the main context, that is cancelled in `Stop` method)

File: pkg/promtail/targets/syslogtarget_test.go [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+type ClientMessage struct {
+	Labels    model.LabelSet
+	Timestamp time.Time
+	Message   string
+}
+
+type TestLabeledClient struct {
+	log      log.Logger
+	messages []ClientMessage
+	sync.RWMutex
```
Comment (pstibrany) [inline pkg/promtail/targets/syslogtarget_test.go]: If this wasn't a test code, embedding mutex like this would be an antipattern.

File: pkg/promtail/targets/syslogparser/syslogparser.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+	defer close(results)
+
+	callback := func(res *syslog.Result) {
+		results <- res
+	}
+
+	octetcounting.NewParser(syslog.WithListener(callback)).Parse(r)
+}
+
+func newlineSeparated(r io.Reader, results chan<- *syslog.Result) {
```
Comment (pstibrany) [inline pkg/promtail/targets/syslogparser/syslogparser.go]: Cannot this be replaced with `nontransparent.NewParser(syslog.WithListener(callback)).Parse(r)`?

**Comment by pstibrany** [comment]
@bastjan btw, if you currently don't have time to take another look at this PR, I have some capacity now and can take over and polish it, if you want.

**Comment by bastjan** [comment]
> @bastjan btw, if you currently don't have time to take another look at this PR, I have some capacity now and can take over and polish it, if you want.

Thanks for the review!

I've got time to polish the PR, should be able to finish it by today. :)

**Comment by pstibrany** [comment]
> I've got time to polish the PR, should be able to finish it by today. :)

That's great, thank you! (No need to hurry :) )

File: pkg/promtail/targets/syslogparser/syslogparser.go (hunk 5 above)
Comment (bastjan) [reply pkg/promtail/targets/syslogparser/syslogparser.go]: Yes, it should be possible. [Telegraf](https://github.com/influxdata/telegraf/blob/1f5be2bac7b895867411406b2eeac892b9ff9258/plugins/inputs/syslog/syslog.go#L322) uses it like this.

However its buffering is a bit strange. The line is emitted only after the next line is received, which is not optimal for connections with very few logs.

File: docs/clients/promtail/configuration.md [hunk 6]
```diff
@@ -580,6 +585,45 @@ labels:
...
+# Label map to add to every log message.
+labels:
+  [ <labelname>: <labelvalue> ... ]
+```
+
+#### Available Labels
+
+* `__syslog_connection_ip_address`: The remote IP address.
+* `__syslog_connection_hostname`: The remote hostname.
+* `__syslog_message_severity`: The syslog severity parsed from the message.
```
[the remark this answers is not in the document]
Comment (bastjan) [reply docs/clients/promtail/configuration.md]: Updated the documentation with references to the RFC and possible values. Let me know what you think.

File: docs/clients/promtail/configuration.md [hunk 7]
```diff
@@ -580,6 +585,45 @@ labels:
...
+* `__syslog_message_facility`: The syslog facility parsed from the message.
+* `__syslog_message_hostname`: The hostname parsed from the message.
+* `__syslog_message_app_name`: The app name parsed from the message.
+* `__syslog_message_proc_id`: The proc id parsed from the message.
+* `__syslog_message_msg_id`: The message id parsed from the message.
+* `__syslog_message_sd_<sd_id>[_<iana_enterprise_id>]_<sd_name>`: The parsed structured data fields from the message.
```
[the remark this answers is not in the document]
Comment (bastjan) [reply docs/clients/promtail/configuration.md]: Updated the documentation with references to the RFC and possible values. Let me know what you think.

File: pkg/promtail/targets/syslogtarget.go [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+	go t.acceptConnections()
+
+	return nil
+}
+
+func (t *SyslogTarget) acceptConnections() {
+	defer t.connectionsClosed.Done()
+
+	l := log.With(t.logger, "address", t.listener.Addr().String())
+	var backoff time.Duration
```
[the remark this answers is not in the document]
Comment (bastjan) [reply pkg/promtail/targets/syslogtarget.go]: Nice! Way more readable...

File: pkg/promtail/targets/syslogtarget.go (hunk 2 above)
Comment (bastjan) [reply pkg/promtail/targets/syslogtarget.go]: The message channel is used to serialize the messages from all connections; this gets rid of out-of-order errors. Every connection copies its syslog messages to this channel. The  [messageSender](https://github.com/grafana/loki/pull/1275/files#diff-9e5d46359c7640bbb963370684c04406R85) ranges over the channel and appends the messages to the EntryHandler.

Since syslog connections can be short-lived only the target knows if every connection is shutdown and if it can close the message channel.

**Comment by bastjan** [comment]
@pstibrany I think I resolved most issues. Take a look and tell me what you think. :)

File: pkg/promtail/targets/syslogtarget.go [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+	defer t.openConnections.Done()
+
+	l := log.With(t.logger, "address", t.listener.Addr().String())
+
+	backoff := util.NewBackoff(t.ctx, util.BackoffConfig{
+		MinBackoff: 5 * time.Millisecond,
+		MaxBackoff: 1 * time.Second,
+	})
+
+	for {
```
Comment (pstibrany) [inline pkg/promtail/targets/syslogtarget.go]:
```suggestion
	for backoff.Ongoing() {
```

File: pkg/promtail/targets/syslogtarget.go [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+		c, err := t.listener.Accept()
+		if err != nil {
+			select {
+			case <-t.ctx.Done():
+				level.Info(l).Log("msg", "syslog server shutting down")
+				return
+			default:
+			}
```
Comment (pstibrany) [inline pkg/promtail/targets/syslogtarget.go]: This can be removed, will be handled by `backoff.Ongoing()` check.

File: pkg/promtail/targets/syslogtarget.go (hunk 2 above)
Comment (pstibrany) [reply pkg/promtail/targets/syslogtarget.go]: Aah, it's the messageSender reacting on close. Got it, thanks!

File: pkg/promtail/targets/syslogtarget_test.go [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+	Labels    model.LabelSet
+	Timestamp time.Time
+	Message   string
+}
+
+type TestLabeledClient struct {
+	log log.Logger
+
+	messages    []ClientMessage
+	messagesMtx sync.RWMutex
```
Comment (pstibrany) [inline pkg/promtail/targets/syslogtarget_test.go]: nitpick: normally, mutex field is above fields that it is protecting.

File: pkg/promtail/targets/syslogparser/syslogparser.go (hunk 5 above)
Comment (pstibrany) [reply pkg/promtail/targets/syslogparser/syslogparser.go]: OK. I wouldn't worry about it too much, hopefully nobody will use this mode, but let's keep it.

File: pkg/promtail/targets/syslogparser/syslogparser.go [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+
+	"github.com/influxdata/go-syslog/v2"
+	"github.com/influxdata/go-syslog/v2/octetcounting"
+	"github.com/influxdata/go-syslog/v2/rfc5424"
+)
+
+// ParseStream parses a rfc5424 syslog stream from the given Reader, returning a channel
+// emitting the parsed messages. The parser automatically detects octet counting.
+// The channel is closed on EOF or unrecoverable errors.
+func ParseStream(r io.Reader) <-chan *syslog.Result {
```
Comment (pstibrany) [inline pkg/promtail/targets/syslogparser/syslogparser.go]: Instead of creating channel and goroutine, can we instead make use this signature:

```suggestion
func ParseStream(r io.Reader, callback func(res *syslog.Result)) error {
```

and pass all parsed messages directly to `callback` function? I think we don't need extra goroutine here, as there is already goroutine for this connection, which is currently just passing messages from one channel to another. It can do the parsing as well.

File: pkg/promtail/targets/syslogtarget.go [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+	}
+}
+
+func (t *SyslogTarget) handleMessageError(err error) {
+	var ne net.Error
+	if errors.As(err, &ne) && ne.Timeout() {
+		level.Debug(t.logger).Log("msg", "connection timed out", "err", ne)
+		return
+	}
+	level.Debug(t.logger).Log("msg", "error parsing syslog stream", "err", err)
```
Comment (pstibrany) [inline pkg/promtail/targets/syslogtarget.go]: This should be at least a warning, so that it's not hidden if someone isn't using debug log.

File: pkg/promtail/targets/syslogparser/syslogparser.go (hunk 12 above)
Comment (bastjan) [reply pkg/promtail/targets/syslogparser/syslogparser.go]: I totally agree! Done :)

File: pkg/promtail/targets/syslogtarget.go (hunk 9 above)
Comment (bastjan) [reply pkg/promtail/targets/syslogtarget.go]: This leads to the ugly "use of closed network connection" error. I check for it inside the loop now: https://github.com/grafana/loki/pull/1275/files#diff-9e5d46359c7640bbb963370684c04406R119

Comment (pstibrany) [reply pkg/promtail/targets/syslogtarget.go]: Ok, makes sense. In that case I'd suggest that we use `if t.ctx.Err() != nil {...}` instead of `if !backoff.Ongoing()` check., since we normally use `backoff.Ongoing()` with for loops.

File: pkg/promtail/targets/syslogparser/syslogparser_test.go [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+
+	require.Equal(t, 2, len(results))
+	require.NoError(t, results[0].Error)
+	require.Equal(t, "First", *results[0].Message.Message())
+	require.NoError(t, results[1].Error)
+	require.Equal(t, "Second", *results[1].Message.Message())
+}
+
+func TestParseStream_NewlineSeparated(t *testing.T) {
+	r := strings.NewReader("<13>1 - - - - - - First\n<13>1 - - - - - - Second")
```
Comment (pstibrany) [inline pkg/promtail/targets/syslogparser/syslogparser_test.go]: Actually, I think that this test is wrong. It may seem like a good idea to use that last unfinished line, but the problem is that we don't know if it is a full message or not, as there is no frame separator.  Perhaps message was cut off in the middle? It would be better to log error in such case.

It also brings us back to using `nontransparent.Parser` – it would simplify the code, and do the right thing. Also non-transparent modes could use different trailers (newline, zero byte, or even crlf -- not supported by this lib) -- if we are going to have it, we may need to expose trailer type as an option.

Comment (bastjan) [reply pkg/promtail/targets/syslogparser/syslogparser_test.go]: Your totally right on the wrong test.

Nontransparent currently supports only newline and null byte. Which we can implement both with scanner and nontransparent. The skipping of the CR happens inside the [rfc5424 parser](https://github.com/influxdata/go-syslog/blob/1193720dd65bc4bab8d9ec1e0c3b5e2496f4d08f/rfc5424/machine.go.rl#L295).

I'm fine with using nontransparent. I just don't like the long, unexpected delays that can happen if there are few messages per connection.

Do you strongly prefer nontransparent? If so I will implement it with nontransparent.
Should I add configuration option for the trailing character with default newline?

Comment (pstibrany) [reply pkg/promtail/targets/syslogparser/syslogparser_test.go]: I am able to reproduce behaviour you're describing (https://gist.github.com/pstibrany/b5fb4e524f478c2f7cd8915aa08f5dc7) but I think it should be reported and fixed in go-syslog package.

Are you running into this bug yourself?

Comment (bastjan) [reply pkg/promtail/targets/syslogparser/syslogparser_test.go]: > Are you running into this bug yourself?

Yes that's the bug.

Comment (pstibrany) [reply pkg/promtail/targets/syslogparser/syslogparser_test.go]: What I mean was whether it affects your deployment. Could you switch to octet-counting?

After some digging, they actually seem to be doing that to handle multiline messages in syslog, see updated Gist.

Comment (bastjan) [reply pkg/promtail/targets/syslogparser/syslogparser_test.go]: > What I mean was whether it affects your deployment. Could you switch to octet-counting?

No it does not, we use syslog-ng where the default is octet-counting. Rsyslog on the other hand uses nontransparent by default, but it is easily configurable. Maybe we should include sample configuration for both?

> After some digging, they actually seem to be doing that to handle multiline messages in syslog, see updated Gist.

Yeah looks like it. Most tools I know don't try to handle newlines in syslog. In rsyslog

```
<39>1 - - - - - - First
another line
<39>1 - - - - - - First
```
becomes
```
<39>1 - - - - - - First
<13>1 - another line - - - 
<39>1 - - - - -B - First
```

Comment (pstibrany) [reply pkg/promtail/targets/syslogparser/syslogparser_test.go]: > Maybe we should include sample configuration for both?

That sounds like great idea!

It's funny that rsyslog doesn't handle multiline messages, as it is in [their documentation](https://www.rsyslog.com/doc/v8-stable/configuration/modules/omfwd.html#tcp-framing) where I found this being mentioned as a possibility.

Comment (bastjan) [reply pkg/promtail/targets/syslogparser/syslogparser_test.go]: :+1:, I will:
- [ ] Use `nontransparent`
- [ ] Document why lines can be delayed.
- [ ] Include sample configuration with octet-counting for syslog-ng and rsyslog under the `syslog_config` section.

Ok?

Comment (pstibrany) [reply pkg/promtail/targets/syslogparser/syslogparser_test.go]: > 👍, I will:
> 
> * [ ]  Use `nontransparent`
> * [ ]  Document why lines can be delayed.
> * [ ]  Include sample configuration with octet-counting for syslog-ng and rsyslog under the `syslog_config` section.
> 
> Ok?

Perfect, thank you!

File: pkg/promtail/targets/syslogparser/syslogparser.go [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+	return nil
+}
+
+func octetCounting(r io.Reader, callback func(res *syslog.Result)) {
+	octetcounting.NewParser(syslog.WithListener(callback)).Parse(r)
+}
+
+func newlineSeparated(r io.Reader, callback func(res *syslog.Result)) {
+	nontransparent.NewParser(syslog.WithListener(callback)).Parse(r)
+}
```
Comment (pstibrany) [inline pkg/promtail/targets/syslogparser/syslogparser.go]: I think these don't need to be methods anymore. Can we inline them?

File: docs/clients/promtail/configuration.md [hunk 16]
```diff
@@ -580,6 +585,68 @@ labels:
...
+### syslog_config
+
+The `syslog_config` block configures a syslog listener allowing users to push
+logs to promtail with the syslog protocol.
+Currently supported is [IETF Syslog (RFC5424)](https://tools.ietf.org/html/rfc5424)
+with and without octet counting.
+
+The recommended deployment is to have a dedicated syslog forwarder like **syslog-ng** or **rsyslog**
+in front of promtail. The forwarder can take care of the various specifications
+and transports that exist (UDP, BSD syslog, ...).
```
Comment (pstibrany) [inline docs/clients/promtail/configuration.md]: ❤️

File: docs/clients/promtail/configuration.md [hunk 17]
```diff
@@ -580,6 +585,68 @@ labels:
...
+# A structured data entry of [example@99999 test="yes"] would become
+# the label "__syslog_message_sd_example_99999_test" with the value "yes".
+label_structured_data: <bool>
+
+# Label map to add to every log message.
+labels:
+  [ <labelname>: <labelvalue> ... ]
+```
+
+#### Syslog-NG Output Configuration
```
Comment (rfratto) [inline docs/clients/promtail/configuration.md]: I wonder if some of these examples should be moved (or copied) to `docs/clients/promtail/scraping.md`. The configuration document has mostly focused on the schema of the YAML file rather than the specifics of configuring it with other components (e.g., syslog-ng, rsyslog).  

What do you think?

File: docs/clients/promtail/configuration.md [hunk 18]
```diff
@@ -970,3 +1037,27 @@ scrape_configs:
...
+```yaml
+server:
+  http_listen_port: 9080
+  grpc_listen_port: 0
+
+positions:
+  filename: /tmp/positions.yaml
+
+clients:
+  - url: http://ip_or_hostname_where_loki_runns:3100/loki/api/v1/push
```
Comment (rfratto) [inline docs/clients/promtail/configuration.md]: Very small nit, but can we simplify the path here?

Maybe just `http://loki_address:3100/loki/api/v1/push`?

File: docs/clients/promtail/configuration.md (hunk 17 above)
Comment (bastjan) [reply docs/clients/promtail/configuration.md]: 👍, I'll try to add a section in the style of the `scraping.md` documentation and link it.

Comment (bastjan) [reply docs/clients/promtail/configuration.md]: Take a look and tell me what you think :)

Comment (rfratto) [reply docs/clients/promtail/configuration.md]: I'm happy with that, nice job!

File: pkg/promtail/targets/syslogtarget.go [hunk 19]
```diff
@@ -0,0 +1,327 @@
...
+// Labels returns the set of labels that statically apply to all log entries
+// produced by the SyslogTarget.
+func (t *SyslogTarget) Labels() model.LabelSet {
+	return t.config.Labels
+}
+
+// Details returns target-specific details.
+func (t *SyslogTarget) Details() interface{} {
+	return map[string]string{}
+}
```
Comment (cyriltovena) [inline pkg/promtail/targets/syslogtarget.go]: I would love to get info useful for checking how does the target, I think we should create an issue, have you seen the /targets and /service-discovery page in Promtail?

Comment (bastjan) [reply pkg/promtail/targets/syslogtarget.go]: > I would love to get info useful for checking how does the target, I think we should create an issue, have you seen the /targets and /service-discovery page in Promtail?

Just took a look at those pages and indeed there is some useful information missing from the syslog target.

I think the most obvious candidates would be open connections in `/targets` and the listen address in `/service-discovery`.

Have to dig in a bit further to see how those pages look in different setups...
